src/zenml/zen_server/rbac/utils.py

```diff
@@ -0,0 +1,569 @@
+#  permissions and limitations under the License.
+"""RBAC utility functions."""
+
+from datetime import datetime
+from enum import Enum
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    TypeVar,
+    Union,
+    cast,
+)
+from uuid import UUID
+
+from fastapi import HTTPException
+from pydantic import BaseModel
+
+from zenml.models import (
+    BaseResponse,
+    Page,
+    UserScopedResponse,
+)
+from zenml.models.base_models import BaseResponseModel, UserScopedResponseModel
+from zenml.zen_server.auth import get_auth_context
+from zenml.zen_server.rbac.models import Action, Resource, ResourceType
+from zenml.zen_server.utils import rbac, server_config
+
+AnyOldResponseModel = TypeVar("AnyOldResponseModel", bound=BaseResponseModel)
+AnyNewResponseModel = TypeVar(
+    "AnyNewResponseModel", bound=BaseResponse  # type: ignore[type-arg]
+)
+AnyResponseModel = TypeVar(
+    "AnyResponseModel",
+    bound=Union[BaseResponse, BaseResponseModel],  # type: ignore[type-arg]
+)
+AnyModel = TypeVar("AnyModel", bound=BaseModel)
+
+
+def dehydrate_page(
+    page: Page[AnyResponseModel],
+) -> Page[AnyResponseModel]:
+    """Dehydrate all items of a page.
+
+    Args:
+        page: The page to dehydrate.
+
+    Returns:
+        The page with (potentially) dehydrated items.
+    """
+    if not server_config().rbac_enabled:
+        return page
+
+    auth_context = get_auth_context()
+    assert auth_context
+
+    resource_list = [get_subresources_for_model(item) for item in page.items]
+    resources = set.union(*resource_list) if resource_list else set()
+    permissions = rbac().check_permissions(
+        user=auth_context.user, resources=resources, action=Action.READ
+    )
+
+    new_items = [
+        dehydrate_response_model(item, permissions=permissions)
+        for item in page.items
+    ]
+
+    return page.copy(update={"items": new_items})
+
+
+def dehydrate_response_model(
+    model: AnyModel, permissions: Optional[Dict[Resource, bool]] = None
+) -> AnyModel:
+    """Dehydrate a model if necessary.
+
+    Args:
+        model: The model to dehydrate.
+        permissions: Prefetched permissions that will be used to check whether
+            sub-models will be included in the model or not. If a sub-model
+            refers to a resource which is not included in this dictionary, the
+            permissions will be checked with the RBAC component.
+
+    Returns:
+        The (potentially) dehydrated model.
+    """
+    if not server_config().rbac_enabled:
+        return model
+
+    if not permissions:
+        auth_context = get_auth_context()
+        assert auth_context
+
+        resources = get_subresources_for_model(model)
+        permissions = rbac().check_permissions(
+            user=auth_context.user, resources=resources, action=Action.READ
+        )
+
+    dehydrated_fields = {}
+
+    for field_name in model.__fields__.keys():
```

Comment:
Same here: should you exclude private fields, just in case a model already uses or decides to use that ?

Reply:
See above, not included